### ISSUE TITLE
Incorrect designation of 'cached' instructions

### ISSUE THREAD

**Issue by rrjjvv** [earthly/earthly#2307]
This is not a new issue; I've verified this behavior in the last five releases, the earliest version I have locally (0.6.5), and some randomly-picked versions in between.  An aspect of the behavior did chance a while back (I'll elaborate at the end).

This is intentionally a little longer than a true minimal reproduction:
```Earthfile
VERSION 0.6

bar:
    FROM alpine
    RUN echo "I'll print"
    # This is the logical failure point for humans, but worse that it's "cached"
    COPY zzz .
    # Second most logical failure point, and extra confusing that it's "cached"
    RUN cat zzz
    # The --if-exists doesn't appear to influence how things fail
    COPY --if-exists xxx .
    # Everything past this point is "cached" up until the very last failure
    RUN echo "anyone?"
    COPY yyy .
    RUN echo "nope"
    COPY xxx .
    RUN --no-cache "I'm cached too..."
    COPY www .
    COPY vvv .
    COPY uuu .
    COPY ttt .
    SAVE IMAGE bar
```

Execution:
```
$ ls -al
total 268
drwxrwxr-x   2 robertovillarreal robertovillarreal   4096 Oct 22 14:16 .
drwxrwxrwt 113 root              root              262144 Oct 22 14:14 ..
-rw-rw-r--   1 robertovillarreal robertovillarreal    612 Oct 22 14:16 Earthfile
<none of the source files used in the copies actually exist>

$ earthly -v
earthly version v0.6.26 4a8d66e3b30d8ee758e7d71b40bc027f92e9adf4 linux/amd64; Ubuntu 20.04.5 LTS (Focal Fossa)

$ docker stop earthly-buildkitd && docker rm -v earthly-buildkitd && docker volume rm earthly-cache
earthly-buildkitd
earthly-buildkitd
earthly-cache

$ docker ps | grep earth; docker volume ls | grep earth
<no output>

$ earthly --no-cache +bar 
 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Starting buildkit daemon as a docker container (earthly-buildkitd)...
           buildkitd | ...Done

 2. Build 🔧
————————————————————————————————————————————————————————————————————————————————

              alpine | --> Load metadata linux/amd64
             context | --> local context .
                +bar | --> FROM alpine
             context | --> local context .
             context | --> local context .
             context | --> local context .
             context | --> local context .
             context | --> local context .
             context | --> local context .
                +bar | [██████████] 100% resolve docker.io/library/alpine@sha256:bc41182d7ef5ffc53a40b044e725193bc10142a1243f395ee852a8d9730fc2ad
                +bar | [██████████] 100% sha256:213ec9aee27d8be045c6a92b7eac22c9a64b44558193775a1a7f626352392b49
                +bar | [██████████] 100% extracting sha256:213ec9aee27d8be045c6a92b7eac22c9a64b44558193775a1a7f626352392b49
                +bar | --> RUN echo "I'll print"
                +bar | I'll print
             context | [          ]   0% transferring .:
             context | [██████████] 100% transferring .:
                +bar | *cached* --> COPY zzz .
                +bar | *cached* --> RUN cat zzz
                +bar | *cached* --> COPY --if-exists xxx .
                +bar | *cached* --> RUN echo "anyone?"
                +bar | *cached* --> COPY yyy .
                +bar | *cached* --> RUN echo "nope"
                +bar | *cached* --> COPY xxx .
                +bar | *cached* --> RUN --no-cache "I'm cached too..."
                +bar | *cached* --> COPY www .
                +bar | *cached* --> COPY vvv .
                +bar | *cached* --> COPY uuu .
                +bar | --> COPY ttt .
                +bar | ERROR Earthfile line 21:4
                +bar |       The command
                +bar |           COPY ttt .
                +bar |       failed: "/ttt": not found
             context | [          ]   0% transferring .:
             context | [          ]   0% transferring .:
             context | [          ]   0% transferring .:
             context | [          ]   0% transferring .:
             context | [          ]   0% transferring .:
             context | [          ]   0% transferring .:
             context | WARN: Canceled

============================ ❌ FAILURE [2. Build 🔧] ============================

Repeating the output of the command that caused the failure
                +bar *failed* | --> COPY ttt .
                +bar *failed* | [no output]
                +bar *failed* | ERROR Earthfile line 21:4
                +bar *failed* |       The command
                +bar *failed* |           COPY ttt .
                +bar *failed* |       failed: "/ttt": not found
Error: File not found "/ttt"
```
I assume that earthly itself doesn't cache to the host filesystem and that it doesn't (can't) utilize the host's (docker) buildkit cache.  So literally deleting the earthly-buildkit container and its cache volume should be sufficient to delete all temporary and permanent caches, thus should be impossible for anything to truly be cached.  My assumption/interpretation is that "\*cached*" is used to generically denote "not executed" (but I'm guessing that "result ignored" is actually closer to the truth).

So the two issues (neither which is new):
1. The output falsely says "\*cached*", when it's clearly not the case.
2. The reported failure is legitimate, but isn't the 'first failure' (from a typical person's perspective)

The combination of the two is what makes this troublesome.  The steps that report as "cached" implies "successfully executed in the past" (when they never did), leaving one to focus on the wrong problem.

---

A minor secondary issue is that there's a built-in "lag" for this type of error.  Running the same build verbosely (truncated to relevant parts):
```
<snip>
                +bar | --> RUN echo "I'll print"
                +bar | I'll print
                +bar | Completed in 211.863173ms
             context | [          ]   0% transferring .:
             context | [██████████] 100% transferring .:
             context | Completed in 5.013408971s
             context | Completed in 0s
                +bar | *cached* --> COPY zzz .
                +bar | Completed in 0s
                +bar | *cached* --> RUN cat zzz
                +bar | Completed in 0s
             context | Completed in 0s
                +bar | *cached* --> COPY --if-exists xxx .
                +bar | Completed in 0s
<snip>
Summary of timing information
Note that the times do not include the expansion of commands like BUILD, FROM, COPY (artifact).
             context | () 35.417000729s
              alpine | () 698.819638ms
                +bar | () 230.262862ms
===============================================================
Total       	36.346083229s
Total (real)	5.817746117s
```
My interpretation: 
1. A copy is given a five second time limit (and either this timeout is what's interpreted as "not found", or if the file is known to not exist, it doesn't cancel the timer)
2. The context took 35 seconds.  There are 8 `COPY` commands, one of which is `--if-exists`.  Could be coincidence, but `7*5=35`, making me think that all of the copies were sent to buildkit in parallel (and why I guessed that "\*cached*" actually meant 'result ignored' and not 'step skipped')

This five second delay wasn't always here.  The 0.6.7 release is where this started; my semi-educated guess is it was new upstream functionality (https://github.com/earthly/buildkit/commit/b3e8c63a48ad8c015f5631fc1947945b229b3919) that was brought into the fork.  Versions 0.6.6 and 0.6.5 both build the same Earthfile in less than a second.

**Comment by rrjjvv** [comment]
A few other observations I didn't include above:
1. It doesn't matter if there's a `SAVE` or not (nor what type)
2. Adding a successful artifact copy in the middle behaves no differently (that step displays as being cached; whether it actually executed I don't know)
3. Adding a bad artifact copy in the middle is different.  All of the bad steps preceeding it still show as cached, but the artifact copy will end the build (and immediately, without the five second delay that local copies incur)

**Comment by alexcb** [comment]
thanks for the repro case; however when I run it, it fails on the `COPY ttt`, which seems to be expected since you mentioned these files shouldn't exist:

```Earthfile
Repeating the output of the command that caused the failure
                +bar *failed* | --> COPY ttt .
                +bar *failed* | [no output]
                +bar *failed* | ERROR Earthfile line 21:4
                +bar *failed* |       The command
                +bar *failed* |           COPY ttt .
                +bar *failed* |       failed: "/ttt": not found
```

Did these files exist at one point, and then you removed them?

**Comment by rrjjvv** [comment]
> it fails on the COPY ttt, which seems to be expected since you mentioned these files shouldn't exist

You're correct; the fact that `COPY ttt` failed is expected.  But it's somewhat unexpected that it was reported.  Every single `COPY` (excluding the `--if-exists`) before it was expected to fail (as well as some of the `RUN` commands).

For a bad analogy, it's like I have `if a && b && c && d {}`, where all of those are `false`, and diagnostics tell me that the condition failed because `d==false`.  That statement is indeed true. And it's true that the entire condition should have failed.  But I would expect the diagnostics to tell me the reason for failure was `a==false`, since you'd expect it to be evaluated left-to-right, and in a short-circuiting manner.

> Did these files exist at one point, and then you removed them?

Nope.

I probably should have done a proper "expected behavior" to match my "actual behavior" in my original report, but what I would have expected (truncated/paraphrased) is one of the following (in order of personal preference):
```
             context | [          ]   0% transferring .:
             context | [██████████] 100% transferring .:
                +bar |  --> COPY zzz .
                +bar *failed* | --> COPY zzz .
                +bar *failed* | [no output]
                +bar *failed* | ERROR Earthfile line 7:4
                +bar *failed* |       The command
                +bar *failed* |           COPY zzz .
                +bar *failed* |       failed: "/zzz": not found
```
(where the very first step that should have failed actually does fail, is also the reported failure, and terminates the build)


```
                +bar | *unevaluated* --> COPY zzz .
                +bar | *unevaluated* --> RUN cat zzz
                +bar | *unevaluated* --> COPY --if-exists xxx .
<snip>
                +bar | --> COPY ttt .
                +bar | ERROR Earthfile line 21:4
                +bar |       The command
                +bar |           COPY ttt .
                +bar |       failed: "/ttt": not found
```
(where the existing reported failure remains the same, explained as "evaluation order is an implementation detail", but with the proceeding steps relabeled in a more honest/intuitive way)


```
                +bar | I'll print
             context | [          ]   0% transferring .:
             context | [██████████] 100% transferring .:
                +bar | --> COPY ttt .
                +bar | ERROR Earthfile line 21:4
                +bar |       The command
                +bar |           COPY ttt .
                +bar |       failed: "/ttt": not found
```
(where existing reported failure remains the same, but all the proceeding steps are entirely omitted from the output)

Hopefully that makes sense...

**Comment by alexcb** [comment]
Ah ha, I understand now.

Here's a much smaller reproduction case:

```Earthfile
VERSION 0.6

test:
    FROM alpine
    RUN echo "this will run"
    COPY missing-file-1 . # this should fail
    RUN false
    COPY missing-file-2 . # BUG: earthly displays failed: "/missing-file-2": not found (instead of failing on COPY missing-file-1 .)
```

**Comment by rrjjvv** [comment]
Yeah, I forgot to elaborate more on why it was longer than it needed to be.
1. to help illustrate the secondary bug behavior (around the fact that a bad `COPY` takes five seconds to fail)
2. to see if that five seconds ever became 'additive'

In my example, even though every `COPY` itself had a five second penalty, there was only a five second real-time penalty.  At one point I had 30 bad copies, curious whether that would turn into a ten second real-time penalty (wondering if max parallelism influenced the number of copy commands executed at once, i.e. the first 20 bad copies took five seconds, and the other 10 bad copies took another five seconds).  Even with the number of bad copies exceeding my parallelism setting, it was still just a five second penalty (if you were curious).
